lib/Dialect/ModArith/Conversions/ModArithToArith/ModArithToArith.cpp

```diff
@@ -14,23 +16,208 @@ namespace mod_arith {
 #define GEN_PASS_DEF_MODARITHTOARITH
 #include "lib/Dialect/ModArith/Conversions/ModArithToArith/ModArithToArith.h.inc"
 
-/// Returns a possibly extended modulus necessary to compute the given operation
-/// without overflow.
-template <typename ValueOrOpResult>
-TypedAttr modulusHelper(IntegerAttr mod, ValueOrOpResult op, bool mul = false,
-                        bool reduce = false) {
-  auto width = getElementTypeOrSelf(op).getIntOrFloatBitWidth();
-  auto modWidth = (mod.getValue() - 1).getActiveBits();
-  width = reduce ? width : std::max(width, mul ? 2 * modWidth : modWidth + 1);
+IntegerType convertModArithType(ModArithType type) {
+  APInt modulus = type.getModulus().getValue();
+  return IntegerType::get(type.getContext(), modulus.getBitWidth());
+}
+
+Type convertModArithLikeType(ShapedType type) {
+  if (auto modArithType = llvm::dyn_cast<ModArithType>(type.getElementType())) {
+    return type.cloneWith(type.getShape(), convertModArithType(modArithType));
+  } else {
+    return type;
+  }
```

Comment:
tiny readability nit:

```suggestion
  }
  
  return type;
```